Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120181
* #120118
* #120370
* #120226
* #120070
* #120043
* __->__ #120042

Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `@run_with_both_funcol_impls`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous

Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `@run_with_both_funcol_impls`.

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -126,7 +126,7 @@ std::vector<at::Tensor> all_reduce_coalesced(
     std::string group_name) {
   std::vector<at::Tensor> outputs;
```
```suggestion
  std::vector<at::Tensor> outputs;
  outputs.reserve(inputs.size());
```
While we are here.

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (approved):
sgtm

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Change the .clone() in native funcol's all_reduce to use at::MemoryFormat::Contiguous"


Summary:
While I think it probably makes more sense to only require `all_reduce` input to be non-overlapping and dense, today `ProcessGroupNCCL` requires it to be contiguous. This is also what the `all_reduce` in non-native funcol does.

Also marking a test affected by this with `run_with_both_funcol_impls`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

ConnollyLeon commented:
Hi, @yifuwang . I got a few questions about the implementation of _c10d_functional.all_reduce, please check.

Why do we need a clone here? I think the additional clone operation would bring extra HBM access that affect the performance of all_reduce. Is it because we need to make the operator to return a tensor different from input, so that compiler could understand its data relevance? Could it be in-placed when applying torch.compile?
